traffic_ops/client/v13/cdnfederations.go

```diff
@@ -0,0 +1,73 @@
+/*
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package v13
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/apache/trafficcontrol/lib/go-tc/v13"
+)
+
+/* Internally, the CDNName is only used in the GET method. The CDNName
+ * seems to primarily be used to differentiate between `/federations` and
+ * `/cdns/:name/federations`. Although the behavior is odd, it is kept to
+ * keep the same behavior from perl. */
+
+func (to *Session) CreateCDNFederationByName(f v13.CDNFederation, CDNName string) (*v13.CreateCDNFederationResponse, ReqInf, error) {
```

Comment:
Since these are new client methods, can you please use the Nullable structs in the parameters and responses? I know most of the other client implementations don't follow that pattern, but IMO this is something we need to curb, especially with new client implementations.

The problem is that some fields in the API are optional, and without using the Nullable struct, a client must provide a value for the optional fields (in the case of Federations, it would be "description"). With a Nullable struct, a client could just leave the description field null if they didn't want a description. This also makes it possible to test null values in the API integration tests.

To see what I mean as an example, this is what I did for the Origins endpoint: https://github.com/apache/trafficcontrol/pull/2247/files#diff-c0c1ea4955f245fa2b7861436ec93374R33

Reply:
Is there no case where a non-nullable struct will be used?

Reply:
The API implementation uses the nullable struct, which really only leaves the client implementation. If the client also uses the nullable struct, then there shouldn't be any need for the non-nullable struct.

A lot of the legacy TO Go client code was already using the non-nullable structs, which is why they're still prevalent. IMO we need to go through all the existing client implementations to add methods that use the Nullable structs and properly decode the responses everywhere. We would then deprecate the existing methods that use non-nullable structs and eventually remove them. But for now we should just prevent adding onto that pile of non-nullable struct implementations.

Reply:
This is outdated, but I didn't change the direct lines of code. Can you close this manually, or will I need to add a small change to what you selected?

Reply:
Don't worry about getting this thread to collapse into the "Show outdated" thing; you can just reply "done" so everyone knows you addressed the comment.